File: DFS/464.Can_I_Win/464.Can_I_Win.cpp
class Solution {
    unordered_map<int, bool> Memo;
public:
    bool canIWin(int maxChoosableInteger, int desiredTotal) {
        if (maxChoosableInteger >= desiredTotal) 
            return true;
        if (2 * desiredTotal > maxChoosableInteger * (1 + maxChoosableInteger)) 
            return false;
        return DFS_with_Memo(maxChoosableInteger, desiredTotal, 0);
    }
    
    bool DFS_with_Memo(int &max, int desired, int status) {
        if (desired <= 0) return false;
        
        if (Memo.count(status)) return Memo[status];
        
        for (int i = 1; i <= max; i++) {
            if ((status >> i) & 1 == 1) continue;
            if (DFS_with_Memo(max, desired - i, status | 1 << i) == false){
                Memo[status] = true;
                return true;
            }
        }
        Memo[status] = false;
        return false;
    }
};

File: DFS/464.Can_I_Win/464.Can_I_Win.py
class Solution:
    def canIWin(self, maxChoosableInteger: 'int', desiredTotal: 'int') -> 'bool':
        if maxChoosableInteger >= desiredTotal:
            return True
        if maxChoosableInteger * (maxChoosableInteger + 1) < 2 * desiredTotal:
            return False
        
        return self.DFS_with_Memo(maxChoosableInteger, desiredTotal, 0, {})
    
    def DFS_with_Memo(self, maxint, desired, status, Memo):
        if desired <= 0:
            return False
        
        if status in Memo:
            return Memo[status]
        
        for i in range(1, maxint + 1):
            if (status >> i) & 1 == 1:
                continue
            if self.DFS_with_Memo(maxint, desired - i, status | (1 << i), Memo) == False:
                Memo[status] = True
                return True;
        Memo[status] = False
        return False